src/main/java/org/isf/utils/jobjects/PaginatedTableDecorator.java

```diff
@@ -0,0 +1,372 @@
+
+			getJLabelCurrentPage();
+			togglePreviousPageButton();
+			toggleNextPageButton();
+		}
+
+		if (paginatorType == PAGELINK || paginatorType == PAGEBROWSER_PAGELINK) {
+
+			pageLinkPanel.removeAll();
+			int totalRows = dataProvider.getTotalRowCount();
+			int pages = (int) Math.ceil((double) totalRows / currentPageSize);
+			ButtonGroup buttonGroup = new ButtonGroup();
+			if (pages > MaxPagingCompToShow) {
+				addPageButton(pageLinkPanel, buttonGroup, 1);
+				if (currentPage > (pages - ((MaxPagingCompToShow + 1) / 2))) {
+					// case: 1 ... n->lastPage
+					pageLinkPanel.add(createEllipsesComponent());
+					addPageButtonRange(pageLinkPanel, buttonGroup, pages - MaxPagingCompToShow + 3, pages);
+				} else if (currentPage <= (MaxPagingCompToShow + 1) / 2) {
+					// case: 1->n ...lastPage
+					addPageButtonRange(pageLinkPanel, buttonGroup, 2, MaxPagingCompToShow - 2);
+					pageLinkPanel.add(createEllipsesComponent());
+					addPageButton(pageLinkPanel, buttonGroup, pages);
+				} else {// case: 1 .. x->n .. lastPage
+					pageLinkPanel.add(createEllipsesComponent());// first ellipses
+					// currentPage is approx mid point among total max-4 center links
+					int start = currentPage - (MaxPagingCompToShow - 4) / 2;
+					int end = start + MaxPagingCompToShow - 5;
+					addPageButtonRange(pageLinkPanel, buttonGroup, start, end);
+					pageLinkPanel.add(createEllipsesComponent());// last ellipsis
+					addPageButton(pageLinkPanel, buttonGroup, pages);// last page link
+				}
+			} else {
+				addPageButtonRange(pageLinkPanel, buttonGroup, 1, pages);
+			}
+			pageLinkPanel.getParent().validate();
+			pageLinkPanel.getParent().repaint();
+		}
+	}
+
+	private Component createEllipsesComponent() {
+		return new JLabel(ELLIPSES, SwingConstants.CENTER);
+	}
+
+	private void addPageButtonRange(JPanel parentPanel, ButtonGroup buttonGroup, int start, int end) {
+		for (; start <= end; start++) {
+			addPageButton(parentPanel, buttonGroup, start);
+		}
+	}
+
+	private void addPageButton(JPanel parentPanel, ButtonGroup buttonGroup, int pageNumber) {
+		JToggleButton toggleButton = new JToggleButton(Integer.toString(pageNumber));
+		toggleButton.setMargin(new Insets(1, 3, 1, 3));
+		buttonGroup.add(toggleButton);
+		parentPanel.add(toggleButton);
+		if (pageNumber == currentPage) {
+			toggleButton.setSelected(true);
+		}
+		toggleButton.addActionListener(ae -> {
+			currentPage = Integer.parseInt(ae.getActionCommand());
+			paginate();
+		});
+	}
+
+	public void paginate() {
+		List<T> rows = dataProvider.getRows(currentPage - 1, currentPageSize);
+		lastPage = (dataProvider.getTotalRowCount() / currentPageSize) + 1;
+		objectTableModel.setObjectRows(rows);
+		objectTableModel.fireTableDataChanged();
+	}
+
+	private JLabel getJLabelCurrentPage() {
+		if (currentPageLabel == null) {
+			currentPageLabel = new JLabel();
+		}
+		currentPageLabel.setText("Page " + currentPage + " / " + lastPage);
```

Comment:
Shouldn't `Page` be in the properties file; something like:
`angal.page.range.fmt.txt = Page {0} / {1}` 
and then use the formatter routine?

Reply:
Yep, doing